Python/pystate.c

```diff
@@ -1698,6 +1698,13 @@ PyThreadState_Clear(PyThreadState *tstate)
 
     Py_CLEAR(((_PyThreadStateImpl *)tstate)->asyncio_running_loop);
 
+    struct llist_node *node;
+    // Clear any lingering tasks so that `TaskObj_finalize` doesn't
+    // try to unregister task from a freed list.
+    llist_for_each_safe(node, &((_PyThreadStateImpl *)tstate)->asyncio_tasks_head) {
+        llist_remove(node);
+    }
```

Comment:
If we go with this overall approach, I think you need a global (per-interpreter) linked list of task objects as well. The TaskObj may outlive the thread. Removing them here means that they will be missing from `asyncio.all_tasks()`. Any remaining tasks should be moved to a shared list when the thread is cleared.

Reply:
> If we go with this overall approach, I think you need a global (per-interpreter) linked list of task objects as well. The TaskObj may outlive the thread.

A task can only outlive it's thread if another thread has a reference to it and that breaks the invariant that asyncio tasks of different loops of different threads shouldn't  be shared. asyncio tasks aren't thread safe as such accessing it from other threads breaks that invariant.

See the thread safety note https://docs.python.org/3/library/asyncio-task.html#asyncio.Task

Reply:
Just to be clear, this code is a preventive measure just so that even if the user messes with tasks, it doesn't crash. Before the event loop exits, all the tasks are removed anyways  so any lingering task would be the ones user added manually via `register_task` *again* after the event loop exited we are good in all cases.